platform/linux-generic/odp_ipsec.c

```diff
@@ -58,15 +58,64 @@ int odp_ipsec_capability(odp_ipsec_capability_t *capa)
 }
 
 int odp_ipsec_cipher_capability(odp_cipher_alg_t cipher,
-				odp_crypto_cipher_capability_t capa[], int num)
+				odp_ipsec_cipher_capability_t capa[], int num)
 {
-	return odp_crypto_cipher_capability(cipher, capa, num);
+	int caps = odp_crypto_cipher_capability(cipher, NULL, 0);
+	odp_crypto_cipher_capability_t crypto_capa[caps];
+	uint32_t req_iv_len;
+	int rc, i, out;
+
+	rc = odp_crypto_cipher_capability(cipher, crypto_capa, caps);
+	if (rc <= 0)
+		return rc;
+
+	req_iv_len = _odp_ipsec_cipher_iv_len(cipher);
+	for (i = 0, out = 0; i < rc; i++) {
+		if (crypto_capa[i].iv_len != req_iv_len)
+			continue;
+
+		if (out < num)
+			capa[out].key_len = crypto_capa[i].key_len;
+		out++;
```

Comment:
"capa[out].key_len =" has to be in the same place where you do "out++". Now it's out of if. I would just add break if  "out" reaches "num"

Reply:
No, because this function should count and return amount of capabilities.

Reply:
```
 * @return Number of capability structures for the algorithm. If this is larger
 *         than 'num', only 'num' first structures were output and application
 *         may call the function again with a larger value of 'num'.
```

Reply:
so you did not fill anything to capa[], increased "out" and returned "out". Then application expects "out" elements in capa[] but some of them are empty. Or I'm wrong?

Reply:
i.e. you can not return more then expected.

Reply:
Actually no. Please consider `odp_crypto_cipher_capability()`/`odp_crypto_auth_capability()` code. They return amount of elements to be filled.

Reply:
```
int odp_crypto_auth_capability(odp_auth_alg_t auth,
			       odp_crypto_auth_capability_t dst[], int num_copy)
{
....
	if (num < num_copy)
		num_copy = num;

	memcpy(dst, src, num_copy * size);

	return num;
```
Here logic is ok. Returned requested num_copy or num if it's smaller.

In patch:
```
odp_crypto_auth_capability_t crypto_capa[caps];
for (i = 0, out = 0; i < rc; i++) {
....
		if (out < num)
￼ 			capa[out].key_len = crypto_capa[i].key_len;
￼		out++;
}
return out;
```
You set up key_len for first num elements, leave rest of element uninitialized and return number of possible elements. Or it's too late for today, but I think I'm correct.

odp_crypto_auth_capability() with dst = NULL calls:
memcpy(dst, src, num_copy * size);
Did you tried it with -O0? I guess memcpy(NULL, *ptr, 0) might fail on some conditions.